### ISSUE TITLE
Make Java source root configurable

### ISSUE THREAD

**Issue by cushon** [bazelbuild/bazel#6353]
Bazel hard-codes some conventions about Java source roots:

https://github.com/bazelbuild/bazel/blob/3aedb2f6de80630f88ffb6b60795c44e351a5810/src/main/java/com/google/devtools/build/lib/rules/java/JavaUtil.java#L79

This is used e.g. to infer [`java_test.test_class`](https://docs.bazel.build/versions/master/be/java.html#java_test.test_class) and to relativize [`java_library.resources`](https://docs.bazel.build/versions/master/be/java.html#java_library.resources).

We should make the source root configurable to support other repository layouts.

**Comment by cushon** [comment]
@lberki pointed out offline that the configuration for this would ideally be per-`WORKSPACE`, rather than a top-level flag or part of the toolchain.

**Comment by petroseskinder** [comment]
Thank you for filing this @cushon. Internally at Two Sigma, the change that we have made is simply adding "testsrc" to the list, and then adding a set of tests in the JavaUtilTest.java class that verified that testsrc worked.

**Comment by petroseskinder** [comment]
Thanks for filing this @cushon. Also discussed with @katre, regarding the matter.
Internally at Two Sigma, we have just appended "testsrc", and added the accompanying test cases to JavaUtilTest.java. Without this, we have to explicitly state our test_class in each of our java_test targets.

**Comment by cushon** [comment]
I'm not sure exactly what the per-`WORKSPACE` config mechanism will look, and supporting `testsrc` seems like a good idea in the interim.

Do you want to open a PR for https://github.com/bazelbuild/bazel/pull/6346/commits/2a8ff16828a27029ed6748ef0827963363d91799 and https://github.com/bazelbuild/bazel/pull/6346/commits/db19ea10eb5340814ee2b129eb741dc1dc13e8b4?

**Comment by lberki** [comment]
Yeah... I'm not sure we should have a knob for this in the first place, and I'd rather have a simple and almost correct fix (i.e. the above two commits) than a complex and almost correct one.

**Comment by petroseskinder** [comment]
This scratches my itch, so it is fine with me if this is not tunable.

**Comment by lberki** [comment]
Not in the short term -- I'd love to, but there probably won't be enough time before next summer.

**Comment by RaresI** [comment]
> @lberki pointed out offline that the configuration for this would ideally be per-`WORKSPACE`, rather than a top-level flag or part of the toolchain.

It is fine to have a default per-WORKSPACE, but please consider that many projects are put together under the same workspace, some even from different companies, so it should at least be configurable per-BUILD file. It's fine to have standards, but the configuration should fit the status-quo, not viceversa.
